File: python/problems/leetcode/0236_lowest_common_ancestor_of_a_binary_tree/solution.py
"""LeetCode 236: Lowest Common Ancestor of a Binary Tree."""

from __future__ import annotations

from typing import Optional


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def lowestCommonAncestor(
        self,
        root: Optional[TreeNode],
        p: TreeNode,
        q: TreeNode,
    ) -> Optional[TreeNode]:
        if root is None or root is p or root is q:
            return root
        left = self.lowestCommonAncestor(root.left, p, q)
        right = self.lowestCommonAncestor(root.right, p, q)
        if left is not None and right is not None:
            return root
        return left if left is not None else right


if __name__ == "__main__":
    # [3,5,1,6,2,0,8,null,null,7,4]
    n7 = TreeNode(7)
    n4 = TreeNode(4)
    n6 = TreeNode(6)
    n2 = TreeNode(2, n7, n4)
    n0 = TreeNode(0)
    n8 = TreeNode(8)
    n5 = TreeNode(5, n6, n2)
    n1 = TreeNode(1, n0, n8)
    root = TreeNode(3, n5, n1)
    assert Solution().lowestCommonAncestor(root, n5, n1) is root
    assert Solution().lowestCommonAncestor(root, n5, n4) is n5
    print("0236 OK")

File: cpp/problems/leetcode/0236_lowest_common_ancestor_of_a_binary_tree/solution.cpp
// LeetCode 236
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
    TreeNode(int v, TreeNode* l, TreeNode* r) : val(v), left(l), right(r) {}
};

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (!root || root == p || root == q) return root;
        TreeNode* l = lowestCommonAncestor(root->left, p, q);
        TreeNode* r = lowestCommonAncestor(root->right, p, q);
        if (l && r) return root;
        return l ? l : r;
    }
};

int main() {
    TreeNode n7(7), n4(4);
    TreeNode n6(6);
    TreeNode n2(2, &n7, &n4);
    TreeNode n0(0), n8(8);
    TreeNode n5(5, &n6, &n2);
    TreeNode n1(1, &n0, &n8);
    TreeNode root(3, &n5, &n1);
    assert(Solution().lowestCommonAncestor(&root, &n5, &n1) == &root);
    assert(Solution().lowestCommonAncestor(&root, &n5, &n4) == &n5);
    cout << "0236 OK" << endl;
    return 0;
}
